File: internal/webhook/event.go
package webhook

import (
	"crypto/rand"
	"encoding/hex"
	"time"

	"github.com/vexarnetwork/vexpay/internal/invoice"
)

// Event is the JSON body delivered to a merchant endpoint.
type Event struct {
	ID      string       `json:"id"`
	Type    string       `json:"type"`
	Created time.Time    `json:"created"`
	Data    invoice.View `json:"data"`
}

// NewEvent builds an event for an invoice status change, e.g. "invoice.paid".
func NewEvent(change invoice.StatusChange, now time.Time) Event {
	return Event{
		ID:      newEventID(),
		Type:    "invoice." + string(change.Invoice.Status),
		Created: now,
		Data:    invoice.NewView(change.Invoice),
	}
}

func newEventID() string {
	var b [12]byte
	if _, err := rand.Read(b[:]); err != nil {
		panic("vexpay: crypto/rand unavailable: " + err.Error())
	}
	return "evt_" + hex.EncodeToString(b[:])
}

File: sdks/php/src/Event.php
<?php

declare(strict_types=1);

namespace VexPay;

/**
 * A webhook event delivered by VexPay, e.g. "invoice.paid".
 */
final class Event
{
    /**
     * @param array<string,mixed> $raw
     */
    public function __construct(
        public readonly string $id,
        public readonly string $type,
        public readonly ?string $created,
        public readonly Invoice $invoice,
        public readonly array $raw,
    ) {
    }

    /**
     * @param array<string,mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: (string) ($data['id'] ?? ''),
            type: (string) ($data['type'] ?? ''),
            created: isset($data['created']) ? (string) $data['created'] : null,
            invoice: Invoice::fromArray($data['data'] ?? []),
            raw: $data,
        );
    }
}
